File: SnipeMobile/API/LocalCacheStore.swift
import Foundation

// Persisted to disk so the app shows data instantly on launch while it refetches.
struct SnipeDataCacheSnapshot: Codable {
    var assets: [Asset] = []
    var users: [User] = []
    var accessories: [Accessory] = []
    var licenses: [License] = []
    var consumables: [Consumable] = []
    var components: [Component] = []
    var locations: [Location] = []
    var companies: [Company] = []
    var manufacturers: [Manufacturer] = []
    var suppliers: [Supplier] = []
    var statusLabels: [StatusLabel] = []
    var savedAt: TimeInterval = Date().timeIntervalSince1970
}

// JSON file cache, one file per server so switching servers never mixes data.
enum LocalCacheStore {
    private static var directory: URL? {
        guard let base = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
            return nil
        }
        let dir = base.appendingPathComponent("SnipeDataCache", isDirectory: true)
        if !FileManager.default.fileExists(atPath: dir.path) {
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir
    }

    private static func fileURL(key: String) -> URL? {
        directory?.appendingPathComponent("\(key).json")
    }

    // Stable, filename-safe key from a base URL.
    static func key(forBaseURL baseURL: String) -> String {
        let raw = baseURL.isEmpty ? "default" : baseURL
        let allowed = CharacterSet(charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789")
        let safe = String(raw.unicodeScalars.map { allowed.contains($0) ? Character($0) : "_" })
        return safe.isEmpty ? "default" : safe
    }

    static func load(key: String) -> SnipeDataCacheSnapshot? {
        guard let url = fileURL(key: key), let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(SnipeDataCacheSnapshot.self, from: data)
    }

    static func save(_ snapshot: SnipeDataCacheSnapshot, key: String) {
        guard let url = fileURL(key: key) else { return }
        guard let data = try? JSONEncoder().encode(snapshot) else { return }
        try? data.write(to: url, options: .atomic)
    }

    // Wipe all cached files (on logout / data wipe).
    static func clearAll() {
        guard let dir = directory else { return }
        try? FileManager.default.removeItem(at: dir)
    }
}

File: Android/app/src/main/java/com/callandt/snipemobile/data/cache/LocalCacheStore.kt
package com.callandt.snipemobile.data.cache

import android.content.Context
import com.callandt.snipemobile.data.model.SnipeDataCacheSnapshot
import com.callandt.snipemobile.data.model.SnipeJson
import java.io.File
import kotlinx.serialization.encodeToString

/** Local JSON cache per server. */
object LocalCacheStore {
    private const val CACHE_DIR = "SnipeDataCache"

    private fun directory(context: Context): File {
        val dir = File(context.filesDir, CACHE_DIR)
        if (!dir.exists()) {
            dir.mkdirs()
        }
        return dir
    }

    private fun file(context: Context, key: String): File =
        File(directory(context), "$key.json")

    /** Stable, filename-safe key from a base URL. */
    fun keyForBaseURL(baseURL: String): String = keyForBaseUrl(baseURL)

    fun keyForBaseUrl(baseURL: String): String {
        val raw = baseURL.ifEmpty { "default" }
        val allowed = ('A'..'Z') + ('a'..'z') + ('0'..'9')
        val safe = raw.map { ch -> if (ch in allowed) ch else '_' }.joinToString("")
        return safe.ifEmpty { "default" }
    }

    fun load(context: Context, key: String): SnipeDataCacheSnapshot? {
        val file = file(context, key)
        if (!file.exists()) return null
        return runCatching {
            SnipeJson.decodeFromString(SnipeDataCacheSnapshot.serializer(), file.readText())
        }.getOrNull()
    }

    fun save(context: Context, snapshot: SnipeDataCacheSnapshot, key: String) {
        val file = file(context, key)
        runCatching {
            val data = SnipeJson.encodeToString(snapshot)
            file.writeText(data)
        }
    }

    /** Wipe all cached files (on data wipe). */
    fun clearAll(context: Context) {
        val dir = directory(context)
        dir.listFiles()?.forEach { it.delete() }
        if (dir.exists()) {
            dir.delete()
        }
    }
}
